plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1238 @@
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* the user has to do something */
+	msg = g_strdup_printf("%s needs to be manually restarted to complete the update. "
+			      "Please unplug the dongle and then re-plug it.",
+			      fu_device_get_name(device));
+	request = fwupd_request_new();
+	fwupd_request_set_kind(request, FWUPD_REQUEST_KIND_IMMEDIATE);
+	fwupd_request_set_id(request, FWUPD_REQUEST_ID_REMOVE_REPLUG);
+	fwupd_request_set_message(request, msg);
+	fu_device_emit_request(device, request);
+
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_prepare(FuDevice *device,
+			     FuProgress *progress,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	SteelseriesSonicWirelessStatus wl_status;
+	guint16 bat_state;
+
+	if (!fu_steelseries_sonic_wireless_status(device, &wl_status, error)) {
+		g_prefix_error(error, "failed to get wireless status: ");
+		return FALSE;
+	}
+	g_debug("WirelessStatus: %u", wl_status);
+	if (wl_status != STEELSERIES_SONIC_WIRELESS_STATE_CONNECTED) {
+		g_autoptr(FwupdRequest) request = NULL;
+		g_autoptr(GTimer) timer = NULL;
+		g_autofree gchar *msg = NULL;
+
+		/* the user has to do something */
+		msg = g_strdup_printf("%s needs to be connected to start the update. "
+				      "Please press any button.",
+				      fu_device_get_name(device));
+		request = fwupd_request_new();
+		fwupd_request_set_kind(request, FWUPD_REQUEST_KIND_IMMEDIATE);
+		fwupd_request_set_id(request, FWUPD_REQUEST_ID_PRESS_UNLOCK);
+		fwupd_request_set_message(request, msg);
+		fu_device_emit_request(device, request);
+
+		/* poll for the wireless status */
+		timer = g_timer_new();
+		do {
+			g_usleep(G_USEC_PER_SEC);
+			if (!fu_steelseries_sonic_wireless_status(device, &wl_status, error)) {
+				g_prefix_error(error, "failed to get wireless status: ");
+				return FALSE;
+			}
+			g_debug("WirelessStatus: %u", wl_status);
+		} while (wl_status != STEELSERIES_SONIC_WIRELESS_STATE_CONNECTED &&
+			 g_timer_elapsed(timer, NULL) * 1000.f <
+			     FU_DEVICE_REMOVE_DELAY_USER_REPLUG);
+		if (wl_status != STEELSERIES_SONIC_WIRELESS_STATE_CONNECTED) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_NEEDS_USER_ACTION,
+					    fu_device_get_update_message(device));
+			return FALSE;
+		}
+	}
+
+	if (!fu_steelseries_sonic_battery_state(device, &bat_state, error)) {
+		g_prefix_error(error, "failed to get battery state: ");
+		return FALSE;
+	}
+	g_debug("BatteryState: %u%%", bat_state);
+	fu_device_set_battery_level(device, bat_state);
+
+	/* success */
+	return TRUE;
+}
+
+static GBytes *
+fu_steelseries_sonic_dump_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	SteelseriesSonicChip chip;
+	gsize blobsz;
+	guint32 bufsz;
+	guint8 *buf;
+	g_autofree guint8 *blob = NULL;
+
+	blobsz = fu_device_get_firmware_size_max(FU_DEVICE(device));
+	blob = g_malloc0(blobsz);
```

Comment:
this feels like the kind of thing you want to use `GByteArray` for.

Reply:
I gave a try and I failed.

How can I auto free the array and it's content?

Reply:
g_autoptr(GByteArray) buf = g_byte_array_new();

and then you can convert to a GBytes using `g_byte_array_free_to_bytes(g_steal_pointer(&buf))`